Implement `volatile_composites` lint

jsgf opened the pull request:
Volatile reads and writes to non-primitive types are not well-defined, and can cause problems.

See https://github.com/rust-lang/rust-clippy/issues/15529 for more details.

Thank you for making Clippy better!

We're collecting our changelog from pull request descriptions.
If your PR only includes internal changes, you can just write
`changelog: none`. Otherwise, please write a short comment
explaining your change.

It's also helpful for us that the lint name is put within backticks (`` ` ` ``),
and then encapsulated by square brackets (`[]`), for example:
```
changelog: [`volatile_composites`]: Lint when read/write_volatile is used on composite types
(structs, arrays, etc) as their semantics are not well defined. Fixes #15529.
```

If your PR fixes an issue, you can add `fixes #issue_number` into this
PR description. This way the issue will be automatically closed when
your PR is merged.

If you added a new lint, here's a checklist for things that will be
checked during review or continuous integration.

- \[ ] Followed [lint naming conventions][lint_naming]
- \[ ] Added passing UI tests (including committed `.stderr` file)
- \[ ] `cargo test` passes locally
- \[ ] Executed `cargo dev update_lints`
- \[ ] Added lint documentation
- \[ ] Run `cargo dev fmt`

[lint_naming]: https://rust-lang.github.io/rfcs/0344-conventions-galore.html#lints

Note that you can skip the above if you are just opening a WIP PR in
order to get feedback.

Delete this line and everything above before opening your PR.

Note that we are currently not taking in new PRs that add new lints. We are in a
feature freeze. Check out the book for more information. If you open a
feature-adding pull request, its review will be delayed.

---

*Please write a short comment explaining your change (or "none" for internal only changes)*

changelog:

samueltardieu commented on clippy_lints/src/volatile_composites.rs:
```diff
@@ -0,0 +1,203 @@
+
+// Make sure the raw pointer has no metadata
+fn is_narrow_raw_ptr<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+    if let ty::RawPtr(_inner, _) = ty.kind() {
+        ty.pointee_metadata_ty_or_projection(cx.tcx).is_unit()
+    } else {
+        false
+    }
```
Are there cases of real `()` metadata? Won't the following be equivalent?

```rust
    matches!(ty.kind(), ty::RawPtr(inner, _) if inner.has_trivial_sizedness(cx.tcx, SizedTraitKind::Sized))
```

samueltardieu commented on clippy_lints/src/volatile_composites.rs:
```diff
@@ -0,0 +1,203 @@
+
+// Enum with some fixed representation and no data-carrying variants
+fn is_enum_repr_c<'tcx>(_cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+    if let ty::Adt(adt_def, _args) = ty.kind()
+        && adt_def.is_enum()
+        && adt_def.repr().inhibit_struct_field_reordering()
+    {
+        adt_def.is_payloadfree()
+    } else {
+        false
+    }
```
Nit: I prefer the shorter equivalent form but it is a matter of  taste.


```suggestion
    ty.ty_adt_def().is_some_and(|adt_def| {
        adt_def.is_enum() && adt_def.repr().inhibit_struct_field_reordering() && adt_def.is_payloadfree()
    })
```

samueltardieu commented on clippy_lints/src/volatile_composites.rs:
```diff
@@ -0,0 +1,203 @@
+
+// We can't know about a generic type, so just let it pass to avoid noise
+fn is_generic<'tcx>(_cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+    ty.flags().intersects(ty::TypeFlags::HAS_PARAM)
```
`ty.flags()` will be removed at some point. Can't you use `ty.has_param()`? Or `ty.has_non_region_param()` if a generic lifetime wouldn't change the lint?

Also, this function is so short that you could probably remove it and replace its call by its content.

samueltardieu commented on clippy_lints/src/volatile_composites.rs:
```diff
@@ -0,0 +1,203 @@
+                    // Raw pointers
+                    ty::RawPtr(innerty, _) => report_volatile_safe(cx, expr, *innerty),
+                    // std::ptr::NonNull
+                    ty::Adt(adt_def, args) if cx.tcx.is_diagnostic_item(sym::NonNull, adt_def.did()) => {
```
Nit: we have a utility for this check in `clippy_utils::ty`:

```suggestion
                    _ if is_type_diagnostic_item(cx, self_ty, sym::NonNull) => {
```

samueltardieu commented on clippy_lints/src/volatile_composites.rs:
```diff
@@ -0,0 +1,203 @@
+                        cx.tcx.get_diagnostic_name(def_id),
+                        Some(sym::ptr_read_volatile | sym::ptr_write_volatile)
+                    )
+                    && let ty::RawPtr(ptrty, _) = cx.typeck_results().expr_ty(arg_ptr).kind()
```
Shouldn't you use `expr_ty_adjusted()` here? The lint currently doesn't catch:

```rust
    let mut s = String::from("foo");
    unsafe {
        std::ptr::write_volatile(&mut s, String::from("bar"));
        //~^ volatile_composites
    }
```

while it probably should.

jsgf replied:
Yes, good catch. I wasn't sure what an "adjustment" was for types.

samueltardieu reviewed (changes requested):
Thanks for the contribution. It is in a good shape already, but I have some questions.

Do not hesitate to add extra tests, especially if my comments were wrong, in order to distinguish between the cases you thought of when the code was more complete than what I suggest.

Also, I would like to have more doc comments on functions, so that it is easier to read.

jsgf commented:
@samueltardieu Thanks very much for your feedback. I'll address your comments shortly. This is the first time I've delved into clippy's internals so I was drawing on existing examples and guesswork for a lot of this change, so your pointers are much appreciated.

jsgf commented on clippy_lints/src/volatile_composites.rs:
```diff
@@ -0,0 +1,203 @@
+// NonNull::{read_volatile,write_volatile}
+
+// primitive type:
+// unit, [iu]{8,16,32,64,128?}, f{32,64}, thin pointer, usize, isize, bool, char
```
Yes, I'll clean these up. In practice this set of types is whatever `Ty::is_primitive` returns true for, which includes `f16` and `f128`.

samueltardieu commented on clippy_lints/src/volatile_composites.rs:
```diff
@@ -0,0 +1,203 @@
+            // Look for method calls to `write_volatile`/`read_volatile`, which
+            // apply to both raw pointers and std::ptr::NonNull.
+            ExprKind::MethodCall(name, self_arg, _, _)
+                if matches!(name.ident.name, sym::read_volatile | sym::write_volatile) =>
```
For the `MethodCall`, since you check for the type and the method name, you know that the right one is called, there is no other possibility. Even if a trait defined a method with the same name and was implemented for raw pointers or `NonNull`, it would not get precedence.

As for `Call`, you match against diagnostic items instead of using function names, so you know that the right ones are called.

samueltardieu commented on clippy_lints/src/volatile_composites.rs:
```diff
@@ -0,0 +1,181 @@
+/// actually generate *any* loads or stores. But this is also used to skip zero-sized
+/// fields of `#[repr(transparent)]` structures.
+fn is_zero_sized_ty<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+    cx.tcx
+        .layout_of(cx.typing_env().as_query_input(ty))
+        .is_ok_and(|layout| layout.is_zst())
```
You can probably further simplify this:

```suggestion
    cx.layout_of(ty).is_ok_and(|layout| layout.is_zst())
```

jsgf replied:
This gets a type error
```
  65 |     cx.tcx.layout_of(ty).is_ok_and(|layout| layout.is_zst())
     |            --------- ^^ expected `PseudoCanonicalInput<'_, Ty<'_>>`, found `Ty<'_>`
```
(there is no `cx.layout_of`)

samueltardieu commented on clippy_lints/src/volatile_composites.rs:
```diff
@@ -0,0 +1,181 @@
+
+/// Top-level predicate for whether a type is volatile-safe or not.
+fn is_volatile_safe_ty<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+    ty.is_primitive()
```
Are all primitive types volatile safe? I would think that there is no chance that 32-bit architectures support atomic read and write to `u128` data for example.

samueltardieu reviewed (changes requested):
This is starting to look good.

I'd like to see two more tests, they may be covered by other paths but it would be more explicit to me:

```rust
struct Empty;

        (&mut Empty as *mut Empty).write_volatile(Empty); // OK
        (0xdead as *mut Wrapper<Empty>).write_volatile(Wrapper((), Empty, ())); // OK
```

Also, regarding primitive types, it looks like this lint should cover non-atomic volatile operations on non-atomic primitive types. Recommending in the description to use primitive types might be misleading, and maybe the name of the lint should be something like `non_atomic_volatile_operations` or something like that.

jsgf commented:
I simplified that ZST predicate and added a test with `Empty`.

> Also, regarding primitive types, it looks like this lint should cover non-atomic volatile operations on non-atomic primitive types. Recommending in the description to use primitive types might be misleading, and maybe the name of the lint should be something like non_atomic_volatile_operations or something like that.

See [my reply](https://github.com/rust-lang/rust-clippy/pull/15686#discussion_r2366431394) to your other comment.

jsgf commented:
(Also remove use of "atomic" in comment in exactly the sense I say we shouldn't be using it.)

samueltardieu reviewed (approved):
Can you please squash the changes?

I hadn't noticed that they were ready to be reviewed because the `S-waiting-on-author` label was set. You can comment with `@rustbot ready` or request a review using GitHub's interface to set the `S-waiting-on-review` label instead.

jsgf commented:
Squashed and rebased to current master.
